pkg/actors/health/health.go

```diff
@@ -0,0 +1,297 @@
+	options := &options{
+		failureThreshold:       defaultFailureThreshold,
+		initialDelay:           defaultInitialDelay,
+		requestTimeout:         defaultRequestTimeout,
+		successStatusCode:      defaultSuccessStatusCode,
+		healthyStateInterval:   defaultHealthyStateInterval,
+		unhealthyStateInterval: defaultUnHealthyStateInterval,
+		clock:                  new(clock.RealClock),
+	}
+
+	for _, o := range opts {
+		o(options)
+	}
+
+	if len(options.address) == 0 {
+		return nil, errors.New("required option 'address' is missing")
+	}
+
+	if options.client == nil {
+		options.client = new(http.Client)
+	}
+
+	return &Checker{
+		ch:                     make(chan bool),
+		client:                 options.client,
+		address:                options.address,
+		initialDelay:           options.initialDelay,
+		requestTimeout:         options.requestTimeout,
+		failureThreshold:       options.failureThreshold,
+		healthyStateInterval:   options.healthyStateInterval,
+		unhealthyStateInterval: options.unhealthyStateInterval,
+		successStatusCode:      options.successStatusCode,
+		clock:                  options.clock,
+		closeCh:                make(chan struct{}),
+	}, nil
+}
+
+func (c *Checker) Run(ctx context.Context) {
+	c.wg.Add(2)
+	defer c.Close()
+	defer c.wg.Done()
+
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		defer c.wg.Done()
+		select {
+		case <-ctx.Done():
+		case <-c.closeCh:
+		}
+		cancel()
+	}()
+
+	select {
+	case <-ctx.Done():
+		return
+	case <-c.clock.After(c.initialDelay):
+	}
+
+	for {
+		c.doUnHealthyStateCheck(ctx)
+		c.doHealthyStateCheck(ctx)
+
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+	}
+}
+
+func (c *Checker) HealthChannel() <-chan bool {
+	return c.ch
+}
+
+func (c *Checker) Close() {
+	defer c.wg.Wait()
+	c.once.Do(func() {
+		close(c.closeCh)
+		c.client.CloseIdleConnections()
+		// Wait for health reports to finish before closing the channel.
+		c.wg.Wait()
+		close(c.ch)
+	})
+}
+
+func (c *Checker) doUnHealthyStateCheck(ctx context.Context) {
+	if ctx.Err() != nil {
+		return
+	}
+
+	ticker := c.clock.NewTicker(c.unhealthyStateInterval)
+	defer ticker.Stop()
+
+	for {
+		// Immediately get health status if in unhealthy state.
+		if c.getStateHealth(ctx) {
+			healthLogger.Info("Actor health check succeeded, marking healthy")
+			c.reportHealth(ctx, true)
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-ticker.C():
+		}
+	}
+}
+
+func (c *Checker) doHealthyStateCheck(ctx context.Context) {
+	if ctx.Err() != nil {
+		return
+	}
+
+	ticker := c.clock.NewTicker(c.healthyStateInterval)
+	defer ticker.Stop()
+
+	var failureCount int32
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-ticker.C():
+			if c.getStateHealth(ctx) {
+				failureCount = 0
+			} else {
+				failureCount++
+			}
+
+			if failureCount >= c.failureThreshold {
+				healthLogger.Warnf("Actor health check failed %d times, marking unhealthy", failureCount)
```

Comment:
This should only be displayed after the first time the actor becomes unhealthy, or it could cause a lot of noise in the logs

Reply:
> This should only be displayed after the first time the actor becomes unhealthy, or it could cause a lot of noise in the logs

A log message is only sent on health state changes, i.e. healthy -> non-healthy and non-healthy to healthy. I think this is a reasonable as a Dapr operator would want to see the last state change from the logs else they would be guessing if they were printed once.